### DISCUSSION TITLE
Mapping issue with query_expression and relationships

### DISCUSSION THREAD

**Discussion by QuentinFchx** [sqlalchemy/sqlalchemy#7200] [Usage Questions] [answered]
Hello,

First of all thank you for this very helpful library.
I'm having an issue with `query_expression/with_expression` and relationships.

I'm using SqlAlch v1.3 + MySQL 5.7

One of my objects has a "complex" property, computed from the state of related objects.
I only want to compute this property in some very specific cases.
From my understanding of the documentation, it looked like `query_expression` was the way-to-go (but that might be wrong 😅).

<details>
  <summary>Click to expand the (not-so-) "simplified" version of the models declaration.</summary>

If needed, I have an even more simplified version of the models that triggers the same error, but I feel like I'm doing something wrong here so I'd like to make sure you have the whole picture.
The problem lies in the `Parent/Child` models. You can probably ignore `Related/Other` and the `hybrid_property`.

```python
import datetime

from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import (
    column_property,
    joinedload,
    query_expression,
    relationship,
    sessionmaker,
    with_expression,
)
from sqlalchemy.sql import case, exists, func, select, text

Base = declarative_base()


class Related(Base):
    __tablename__ = "related"

    id = Column(Integer, primary_key=True)

    checkin = Column(DateTime, index=True)
    checkout = Column(DateTime, index=True)

    parent_id = Column(Integer, ForeignKey("parent.id"), nullable=True, unique=True)
    parent = relationship("Parent", back_populates="related")

    child_id = Column(Integer, ForeignKey("child.id"), nullable=True, unique=True)
    child = relationship("Child", back_populates="related")


class Other(Base):
    __tablename__ = "other"

    id = Column(Integer, primary_key=True)

    status = Column(String(50), index=True)

    child_id = Column(Integer, ForeignKey("child.id"))
    child = relationship("Child", back_populates="other")


class Parent(Base):
    __tablename__ = "parent"

    id = Column(Integer, primary_key=True)

    related = relationship("Related", uselist=False, back_populates="parent")

    children = relationship("Child")

    @hybrid_property
    def is_passed(self):
        return self.related.checkin <= datetime.datetime.now()

    @is_passed.expression
    def is_passed(cls):
        return case(
            [(Related.checkout.isnot(None), Related.checkout < datetime.datetime.now())],
            else_=func.date_add(Related.checkin, text("interval 12 hour")) < datetime.datetime.now(),
        )


class Child(Base):
    __tablename__ = "child"

    id = Column(Integer, primary_key=True)

    parent_id = Column(Integer, ForeignKey("parent.id"), nullable=False)
    parent = relationship("Parent")

    related = relationship("Related", uselist=False, back_populates="child")
    other = relationship("Other", uselist=False, back_populates="child")

    status = Column(String(50), index=True)

    # Tadaaa 🎉 
    complex_field = query_expression()
```
</details>

What I'm trying to achieve is the following:

I want to get the `Parent` objects, with all its children WITH the `complex_field`

```python
# imports are done in the model declaration above

def inject_child_complex_field(query):
    expr = case(
        [
            (Child.status == "foo", "bar"),
            (
                Child.status == "qux",
                case(
                    [(Parent.is_passed.is_(True), "quz")],
                    else_=case(
                        [
                            (
                                exists(
                                    select([Other])
                                    .where(Other.child_id == id)
                                    .where(Other.status.in_(["other_status_1"]))
                                ),
                                "qop",
                            )
                        ],
                        else_=None,
                    ),
                ),
            ),
        ],
        else_=None,
    )

    query = query.options(with_expression(Child.complex_field, expr))
    return query

engine = create_engine("mysql://root:@localhost/privateaser_dev?charset=utf8mb4")
Session = sessionmaker(bind=engine)
session = Session()

query = session.query(Parent).options(joinedload(Parent.related)).options(joinedload(Parent.children))
query = inject_child_complex_field(query)
parent = query.first()
print(parent.children[0].complex_field) # should not be None
```

The error I get instead:
```
sqlalchemy.exc.ArgumentError: Mapped attribute "Child.complex_field" does not apply to any of the root entities in this query, e.g. mapped class Parent->parent. Please specify the full path from one of the root entities to the target attribute.
```

I am obviously missing something, but I don't know what.
I don't think `hybrid_property` is a solution here (I don't have the need to compute the property in Python).
Should I use a `column_property` instead? But then, how can I add joins statement in a `column_property`?

Thank you in advance for any clues or advices! 😄

**Comment by CaselIT** [comment, marked as answer]
Hi,

Just guessing here, but I think what the error is trying to indicate that you should apply `with_expression` to the `joinedload(Parent.children)`.

Try something like this `query.options(joinedload(Parent.children).with_expression(Child.complex_field, expr))`

**Reply by QuentinFchx** [reply]
Hi, thanks a lot for your help!

It seems to solve the error which is very good! I didn't realize that the `joinedload` expression is itself kind of a Query object.

_However_ the resulting query includes what seems to be a cartesian product on the `Related` object.
Most likely because in the complex field I use the `Parent`'s `hybrid_property` which uses its `Related` object (I know, my usecase is tricky 🙈).

Now that I know I can manipulate the `joinedload` expression, I'll continue to investigate.
I'll most likely accept your answer since it does remove the error, I just want to make sure I get to the bottom of the problem first :) !

<details>
  <summary>Just in case, here is the generated query</summary>

```sql
SELECT parent.id AS parent_id, related_1.id AS related_1_id, related_1.checkin AS related_1_checkin, related_1.checkout AS related_1_checkout, related_1.parent_id AS related_1_parent_id, related_1.child_id AS related_1_child_id,
  CASE WHEN (child_1.`status` = %s) THEN %s WHEN (child_1.`status` = %s) THEN CASE WHEN (CASE WHEN (related.checkout IS NOT NULL) THEN related.checkout < %s ELSE date_add(related.checkin, interval 12 hour) < %s END IS true) THEN %s ELSE CASE WHEN (EXISTS (SELECT other.id, other.`status`, other.child_id
  FROM other
  WHERE other.child_id = %s AND other.`status` IN (%s))) THEN %s END END END AS anon_1,
child_1.id AS child_1_id, child_1.parent_id AS child_1_parent_id, child_1.`status` AS child_1_status
FROM related, parent
LEFT OUTER JOIN related AS related_1 ON parent.id = related_1.parent_id
LEFT OUTER JOIN child AS child_1 ON parent.id = child_1.parent_id
```
</details>

**Reply by CaselIT** [reply]
it's probably best to wait for @zzzeek input here,

**Reply by zzzeek** [reply]
well I think you want to use query.join() / query.select_from(), not joinedload().  joinedload() is going to [create a join that you can't otherwise refer towards](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html?highlight=zen#the-zen-of-joined-eager-loading).  if you are adding column expressions that refer to other FROM elements , you need to use select_from() / join() to ensure they are stated in the FROM clause in the way you want.

**Reply by QuentinFchx** [reply]
Thank you both for your quick answers!
We managed to solve the two issues, the mapping error + the join spaghetti 🍝 !

Plus your helped us realize we were doing premature optimizations on some queries and we could simplify things.

Again, thank you for your time 👍
